lib/codegen/Util.cpp

```diff
@@ -1,13 +1,81 @@
 #include "llvm/Support/raw_ostream.h"
 
 namespace kllvm {
 
+// this is an adaptation of a function of the same name in llvm/lib/IR/Function.cpp
+// that unfortunately was not made public by llvm. We use it to assist with
+// the process of generating function names for "intrinsics" that we need to generate
+// in order to deal with the limitations of the garbage collection API in llvm
+std::string getMangledTypeStr(llvm::Type *Ty, bool &HasUnnamedType) {
+  using namespace llvm;
+  std::string Result;
+  if (PointerType* PTyp = dyn_cast<PointerType>(Ty)) {
+    Result += "p" + std::to_string(PTyp->getAddressSpace()) +
+              getMangledTypeStr(PTyp->getElementType(), HasUnnamedType);
+  } else if (ArrayType* ATyp = dyn_cast<ArrayType>(Ty)) {
+    Result += "a" + std::to_string(ATyp->getNumElements()) +
+              getMangledTypeStr(ATyp->getElementType(), HasUnnamedType);
+  } else if (StructType *STyp = dyn_cast<StructType>(Ty)) {
+    if (!STyp->isLiteral()) {
+      Result += "s_";
+      if (STyp->hasName())
+        Result += STyp->getName();
+      else
+        HasUnnamedType = true;
+    } else {
+      Result += "sl_";
+      for (auto Elem : STyp->elements())
+        Result += getMangledTypeStr(Elem, HasUnnamedType);
+    }
+    // Ensure nested structs are distinguishable.
+    Result += "s";
+  } else if (FunctionType *FT = dyn_cast<FunctionType>(Ty)) {
+    Result += "f_" + getMangledTypeStr(FT->getReturnType(), HasUnnamedType);
+    for (size_t i = 0; i < FT->getNumParams(); i++)
+      Result += getMangledTypeStr(FT->getParamType(i), HasUnnamedType);
+    if (FT->isVarArg())
+      Result += "vararg";
+    // Ensure nested function types are distinguishable.
+    Result += "f";
+  } else if (Ty) {
+    switch (Ty->getTypeID()) {
+    default: llvm_unreachable("Unhandled type");
+    case Type::VoidTyID:      Result += "isVoid";   break;
+    case Type::MetadataTyID:  Result += "Metadata"; break;
+    case Type::HalfTyID:      Result += "f16";      break;
+    case Type::FloatTyID:     Result += "f32";      break;
+    case Type::DoubleTyID:    Result += "f64";      break;
+    case Type::X86_FP80TyID:  Result += "f80";      break;
+    case Type::FP128TyID:     Result += "f128";     break;
+    case Type::PPC_FP128TyID: Result += "ppcf128";  break;
+    case Type::X86_MMXTyID:   Result += "x86mmx";   break;
+    case Type::IntegerTyID:
+      Result += "i" + std::to_string(cast<IntegerType>(Ty)->getBitWidth());
+      break;
+    }
+  }
+  return Result;
+}
+
+llvm::Function* koreHeapAlloc(ValueType Cat, std::string name, llvm::Module *module) {
+  llvm::Type *size_type = llvm::Type::getInt64Ty(module->getContext());
+  llvm::Type *Ty = getParamType(Cat, module);
+  auto allocType = llvm::FunctionType::get(Ty, llvm::ArrayRef<llvm::Type*>(size_type), false);
+  bool hasUnnamed = false;
+  auto result = getOrInsertFunction(module, name + "_" + getMangledTypeStr(Ty, hasUnnamed), allocType);
```

Comment:
Stylistic again - do we have a set of shared code style guidelines for C++? Here we're mixing `auto` and non-`auto` declarations, for example. There isn't a clang-format style set, so perhaps it's worth having that to keep things consistent.

Reply:
We don't really. I haven't used clang-format before, so if you have suggestions about how to set things up so that stylistic things of this nature can be checked automatically, I'm willing to listen, as long as it doesn't involve requiring us to duplicate the types of variables in both the declaration and initializer.

Reply:
I can push a PR demonstrating the use of clang-format; it's responsible for purely aesthetic things, and integrates nicely with pretty much every editor and is very low-maintenance once it's set up.

clang-tidy is a bit more involved and probably suits an occasional pass through the codebase using it to see if anything is obviously wrong / could be modernised. Happy to demonstrate that as well.